File: 809/809_version1.py
class Solution:
    def expressiveWords(self, S: str, words: list) -> int:
        size = len(words)
        if size == 0:
            return size
        elif not S:
            return 0
        else:
            numder = 0
            self.target_list = self.countWord(S)
            for w in words:
                if self.compareWords(w):
                    numder += 1
            return numder

    def countWord(self, word):
        result = []
        start = 0
        while start < len(word):
            end = start + 1
            while end < len(word) and word[end] == word[end - 1]:
                end += 1
            # if end == len(word):
            #     break
            distance = end - start
            result.append((word[start], distance))
            start = end
        return result

    def compareWords(self, source):
        source_list = self.countWord()
        if len(source_list) != len(self.target_list):
            return False
        else:
            index = 0
            while index < len(source_list):
                s_char, s_len, t_char, t_len = source_list[index], self.target_list[index]
                if s_char != t_char:
                    return False
                elif s_len > t_len:
                    return False
                elif s_len == 1 and t_len == 2:
                    return False
                index += 1
            return True


def test():
    a = Solution()
    b = a.countWord("hello")
    print(b)


test()

File: 809/809_version1.java
class Solution {
    private List<Character> targetChar = new ArrayList<>();
    private List<Integer> targetNumber = new ArrayList<>();
    public int expressiveWords(String S, String[] words) {
        int size = words.length;
        if(size==0 || S.length()==0)
            return 0;
        countWords(targetChar,targetNumber,S); /** initialize target word's information*/
        int result = 0;
        for(String word:words){
            if(compareWords(word)) result++;
        }
        return result;
    }
    
    private void countWords(List<Character> charList, List<Integer> numberList,String word){
        if(word.length()==0)
            return;
        else{
            int index = 0;
            while(index < word.length()){
                int end = index+1; /**Two pointers */
                while(end < word.length()&&word.charAt(end)==word.charAt(end-1))
                    end+=1;
                charList.add(word.charAt(index));
                numberList.add(end-index);
                index = end;
            }
        }
    }
    
    private boolean compareWords(String word){
        List<Character> wordChar = new ArrayList<>();
        List<Integer> wordNumber = new ArrayList<>();
        countWords(wordChar,wordNumber,word);
        if(wordChar.size()!=targetChar.size())
            return false;
        int index = 0;
        while(index < wordChar.size()){
            Character tChar = targetChar.get(index);
            Character sChar = wordChar.get(index);
            if(tChar!=sChar) return false;
            Integer tNumber = targetNumber.get(index);
            Integer sNumber = wordNumber.get(index);
            if(tNumber==2 && sNumber==1) return false;
            if(tNumber < sNumber) return false;
            index++;
        }
        return true;
    }
}
